source/extensions/http/header_validators/envoy_default/header_validator.cc

```diff
@@ -434,6 +434,48 @@ bool HeaderValidator::hasChunkedTransferEncoding(const HeaderString& value) {
   return false;
 }
 
+// For both H/1 and H/2 trailers should only contain generic headers. As such a common validation
+// method can be used. For H/2, per https://www.rfc-editor.org/rfc/rfc9113#section-8.1 trailers MUST
```

Comment:
I'm missing where we validate pseudo headers?
Also if we don't remove pseudo headers for HTTP/1.1 will we somewhere where we're doing protocol downgrade?

Reply:
We validate them in [H/1](https://github.com/envoyproxy/envoy/blob/bc3d03886cd4ca8b9677cbaba7976fc2840b9597/source/extensions/http/header_validators/envoy_default/http1_header_validator.cc#L53) and [H/2](https://github.com/envoyproxy/envoy/blob/bc3d03886cd4ca8b9677cbaba7976fc2840b9597/source/extensions/http/header_validators/envoy_default/http2_header_validator.cc#L50) specific validators. Although the function is the same in both cases and should be refactored (I'll file an Issue for this). The `validateGenericHeaderName` fails if there is ':' in the name.

For H/1 we never remove pseudoheaders, since the H/1 parser can not produce them. We only add them for the request and response using the first line from the wire. And we never add them for trailers.

Reply:
ah, the comment reads like pseudo headers are allowed for HTTP/1.1   maybe make that clear?

Reply:
Yes, the comment is confusing. I've tried to clarify it, see if it is better. I'd like to provide more details from RFCs on why the common function would work.